includes/data-events/class-webhooks.php

```diff
@@ -0,0 +1,615 @@
+<?php
+/**
+ * Newspack Data Events Webhooks.
+ *
+ * @package Newspack
+ */
+
+namespace Newspack\Data_Events;
+
+use Newspack\Logger;
+use WP_Error;
+
+/**
+ * Main class.
+ */
+final class Webhooks {
+
+	/**
+	 * Webhook request post type.
+	 */
+	const REQUEST_POST_TYPE = 'np_webhook_request';
+
+	/**
+	 * Webhook endpoint taxonomy.
+	 */
+	const ENDPOINT_TAXONOMY = 'np_webhook_endpoint';
+
+	/**
+	 * Maximum number of request retries before disabling the endpoint.
+	 */
+	const MAX_RETRIES = 15;
+
+	/**
+	 * Time to retain finished requests.
+	 */
+	const DELETE_REQUESTS_BEFORE = '7 days ago';
+
+	/**
+	 * Header to be used while logging.
+	 */
+	const LOGGER_HEADER = 'NEWSPACK-WEBHOOKS';
+
+	/**
+	 * Initialize hooks.
+	 */
+	public static function init() {
+		\add_action( 'init', [ __CLASS__, 'register_request_post_type' ] );
+		\add_action( 'init', [ __CLASS__, 'register_endpoint_taxonomy' ] );
+		\add_action( 'init', [ __CLASS__, 'register_cron_events' ] );
+		\add_action( 'newspack_data_event_dispatch', [ __CLASS__, 'handle_dispatch' ], 10, 4 );
+		\add_action( 'transition_post_status', [ __CLASS__, 'transition_post_status' ], 10, 3 );
+	}
+
+	/**
+	 * Register custom post type for webhook requests
+	 */
+	public static function register_request_post_type() {
+		\register_post_type(
+			self::REQUEST_POST_TYPE,
+			[
+				'labels'       => [
+					'name'          => __( 'Webhook Requests', 'newspack' ),
+					'singular_name' => __( 'Webhook Request', 'newspack' ),
+				],
+				'public'       => false,
+				'show_ui'      => false,
+				'show_in_menu' => false,
+				'show_in_rest' => false,
+				'supports'     => [
+					'title',
+					'custom-fields',
+				],
+				'taxonomies'   => [
+					self::ENDPOINT_TAXONOMY,
+				],
+			]
+		);
+	}
+
+	/**
+	 * Register taxonomy for webhook endpoints
+	 */
+	public static function register_endpoint_taxonomy() {
+		\register_taxonomy(
+			self::ENDPOINT_TAXONOMY,
+			self::REQUEST_POST_TYPE,
+			[
+				'labels'       => [
+					'name'          => __( 'Webhook Endpoints', 'newspack' ),
+					'singular_name' => __( 'Webhook Endpoint', 'newspack' ),
+				],
+				'public'       => false,
+				'show_ui'      => false,
+				'show_in_menu' => false,
+				'show_in_rest' => false,
+			]
+		);
+	}
+
+	/**
+	 * Get cron configuration.
+	 *
+	 * "clear_finished": Twice a day will permanently delete finished webhook
+	 * requests older than 7 days.
+	 *
+	 * "send_late_requests": Hourly will send webhook requests that are late.
+	 *
+	 * @return array Schedules keyed by the class method that should be called.
+	 */
+	private static function get_cron_config() {
+		return [
+			'clear_finished'     => 'twicedaily',
+			'send_late_requests' => 'hourly',
+		];
+	}
+
+	/**
+	 * Register webhook cron events.
+	 */
+	public static function register_cron_events() {
+		\register_deactivation_hook( __FILE__, [ __CLASS__, 'clear_cron_events' ] );
```

Comment:
non blocking: is this the best place for this? I think for deactivation hook it's better something that's always ran when you include the main plugin file, without conditionals. Even if there's nothing to cleanup, I think that's fine

Reply:
Great point! I don't think we can register multiple activation and deactivation hooks like that.

72188207f86c29fc33f786c7a652e544388030f5 creates a centralized hook on the main file. You can test it with:

```php
add_action( 'newspack_activation', function() { error_log( 'activating newspack' ); } );
add_action( 'newspack_deactivation', function() { error_log( 'deactivating newspack' ); } );
```

Deactivate, activate and confirm the logs.